### ISSUE TITLE
'bool' object has no attribute 'parameters'

### ISSUE THREAD

**Issue by Martinjiajun** [cvxpy/cvxpy#1090]
Hi，I am an ordinary graduate student.I encountered a problem when using Cvxpy recently.
[13节点算例.xlsx](https://github.com/cvxgrp/cvxpy/files/4964719/13.xlsx)
```
import cvxopt
import xlrd
import numpy as np
import cvxpy as cvx
import math
Sbase=1000.0
Vbase=4.16/math.sqrt(3)
Xbase=Vbase**2*1e3/Sbase
data = xlrd.open_workbook(r"C:\Users\mjj\13节点算例.xlsx")
sheet1=data.sheet_by_name(u'发电机')
nrows=sheet1.nrows
ncols =sheet1.ncols
gen0=np.zeros((nrows-2,ncols))
for i in range(2,nrows):
    for k in range(2):
        gen0[i-2][k]=sheet1.col(k)[i].value
gen0=gen0.astype(int)
sheet2=data.sheet_by_name(u'负荷')
nrows_2=sheet2.nrows
ncols_2 =sheet2.ncols
bus0=np.zeros((nrows_2-2,ncols_2))
for i in range(2,nrows_2):
    for k in range(8):
        bus0[i-2][k] = sheet2.col(k)[i].value
sheet3=data.sheet_by_name(u'线路')
nrows_3=sheet3.nrows
ncols_3 =sheet3.ncols
branch0=np.zeros((nrows_3-2,ncols_3))`

for i in range(2,nrows_3):
    for k in range(4):
        branch0[i-2][k] = sheet3.col(k)[i].value

bus=bus0
branch=branch0
gen=gen0
n_bus=bus.shape[0] 
n_gen=gen.shape[0] 
n_branch=branch.shape[0] 
w=bus[:,7] 
root=gen[0,0] 
v0=np.array([[1],[-0.5-1j*math.sqrt(3)/2],[-0.5+1j*math.sqrt(3)/2]]) 
V0=v0*(np.conj(v0).T) 
z=[None]*13 
z[0]=np.eye(3)*1e-3 
#601
z[1]=np.array([[0.3465 + 1.0179j,0.1560 + 0.5017j,0.1580 + 0.4236j],
               [0.1560 + 0.5017j,0.3375 + 1.0478j,0.1535 + 0.3849j],
               [0.1580 + 0.4236j,0.1535 + 0.3849j,0.3414 + 1.0348j]])
#602
z[2]=np.array([[0.7526 + 1.1814j,0.1580 + 0.4236j,0.1560 + 0.5017j],
              [0.1580 + 0.4236j,0.7475 + 1.1983j,0.1535 + 0.3849j],
              [0.1560 + 0.5017j,0.1535 + 0.3849j,0.7436 + 1.2112j]])
#603
z[3]=np.array([[0.0000 + 0.0000j,0.0000 + 0.0000j,0.0000 + 0.0000j],
              [0.0000 + 0.0000j,1.3294 + 1.3471j,0.2066 + 0.4591j],
              [0.0000 + 0.0000j,0.2066 + 0.4591j,1.3238 + 1.3569j]])
#1004
z[4]=np.array([[1.3238 + 1.3569j,0.0000 + 0.0000j,0.2066 + 0.4591j],
               [0.0000 + 0.0000j,0.0000 + 0.0000j,0.0000 + 0.0000j],
               [0.2066 + 0.4591j,0.0000 + 0.0000j,1.3294 + 1.3471j]])
#1005
z[5]=np.array([[0.0000 + 0.0000j,0.0000 + 0.0000j,0.0000 + 0.0000j],
               [0.0000 + 0.0000j,0.0000 + 0.0000j,0.0000 + 0.0000j],
               [0.0000 + 0.0000j,0.0000 + 0.0000j,1.3292 + 1.3475j]])
#1006
z[6]=np.array([[0.7982 + 0.4463j,0.3192 + 0.0328j,0.2849 - 0.0143j],
               [0.3192 + 0.0328j,0.7891 + 0.4041j,0.3192 + 0.0328j],
               [0.2849 - 0.0143j,0.3192 + 0.0328j,0.7982 + 0.4463j]])
#1007
z[7]=np.array([[1.3425 + 0.5124j,0.0000 + 0.0000j,0.0000 + 0.0000j],
               [0.0000 + 0.0000j,0.0000 + 0.0000j,0.0000 + 0.0000j],
               [0.0000 + 0.0000j,0.0000 + 0.0000j,0.0000 + 0.0000j]])
z[8]=np.eye(3)*(38.072+69.222j) 
Z=[None]*n_branch 
for i in range(n_branch):
    O=z[int(branch[i,3])]
    Z[i]=O/Xbase*branch[i,2]/5280
Vmax=[1.1025,1.1025,1.1025] 
Vmin=[0.9025,0.9025,0.9025]
Alpha=np.exp(-2*math.pi*1j/3)
Gamma=np.array([[1.0,Alpha*Alpha,Alpha],
               [Alpha,1.0,Alpha*Alpha],
               [Alpha*Alpha,Alpha,1.0]])
A11=np.eye(3)
M=100#一个很大的数
V = {}
for i in range(n_bus):
    V[i] = cvx.Variable((3,3),complex=True) 
Lam={}
for i in range(n_branch):
    Lam[i]=cvx.Variable((3,1),complex=True) 
S={}
for i in range(n_branch):
    S[i]=cvx.Variable((3,3),complex=True)
a = cvx.Variable(n_branch,boolean=True) 
b = cvx.Variable(2*n_branch,boolean=True) 
sw = cvx.Variable(n_bus,boolean=True) 
F=0
for i in range(n_bus):
    F=F+sw[i]*w[i] 
obj=cvx.Maximize(F) 
#设定约束
constraints=[]
for i in range(n_bus):
    if bus[i,7]==0:
        constraints.append(sw[i]==0)
for i in range(n_bus):
    S_to=0
    S_from=0
    for j in range(n_branch):
        if branch[j,0]==i:
            S_to=S_to+Lam[j]            
        if branch[j,1]==i:
            S_from=S_from+Lam[j]
            if Z[j][0,0]==0:
                constraints.append(cvx.real(Lam[j][0,0])==0)
                constraints.append(cvx.imag(Lam[j][0,0])==0)
            if Z[j][1,1]==0:
                constraints.append(cvx.real(Lam[j][1,0])==0)
                constraints.append(cvx.imag(Lam[j][1,0])==0)
            if Z[j][2,2]==0:
                constraints.append(cvx.real(Lam[j][2,0])==0)
                constraints.append(cvx.imag(Lam[j][2,0])==0)   
S_load=np.array([[(bus[i,1]+bus[i,2]*1j)/Sbase],[(bus[i,3]+bus[i,4]*1j)/Sbase],[(bus[i,5]+bus[i,6]*1j)/Sbase]])*sw[i]
fff=1   
for k in range(n_gen):
    if gen[k,0]==i:
        constraints.append(cvx.real(sum((S_to-S_from+S_load)))<=gen[k,1]/Sbase)
        fff=0
if fff==1:
    constraints.append(cvx.real(S_from-S_to)==cvx.real(S_load))
    constraints.append(cvx.imag(S_from-S_to)==cvx.imag(S_load))
for i in range(n):
    if bus[i,0]==0:
        constraints.append(V[i]==V0) 
    else:
        for j in range(3):
            constraints.append(cvx.real(V[i][j,j])>=Vmin[j]) 
            constraints.append(cvx.real(V[i][j,j])<=Vmax[j])
            constraints.append(cvx.imag(V[i][j,j])==0) 
for i in range(n_branch):
    constraints.append(S[i]==Gamma@cvx.diag(Lam[i]))
    constraints.append(cvx.real(V[branch[i,0]]-V[branch[i,1]])<=cvx.real(M*(1-a[i])*A11+S[i]@np.transpose(Z[i])+Z[i]@cvx.transpose(S[i])))
    constraints.append((cvx.real(-M*(1-a[i])*A11+S[i]@np.transpose(Z[i])+Z[i]@cvx.transpose(S[i]))<=cvx.real(V[branch[i,0]]-V[branch[i,1]])))
    constraints.append(cvx.imag(V[branch[i,0]]-V[branch[i,1]])<=M*(1-a[i])*A11+cvx.imag(S[i]@np.transpose(Z[i])+Z[i]@cvx.transpose(S[i])))
    constraints.append(-M*(1-a[i])*A11+cvx.imag(S[i]@np.transpose(Z[i])+Z[i]@cvx.transpose(S[i]))<=cvx.imag(V[branch[i,0]]-V[branch[i,1]]))
    constraints.append(b[2*i-1]+b[2*i]==a[i])   
for i in range(n_bus):
    b_all=0
    flag1=1
    if root==i:
        for j in range(n_branch):
            if branch[j,0]==i:
                constraints.append(b[j*2-1]==0)
                flag1=0
            if branch[j,1]==i:
                constraints.append(b[j*2]==0)
                flag1=0
    else:
        for j in range(n_branch):
            if branch[j,0]==i:
                b_all=b_all+b[2*j-1]
            if branch[j,1]==i:
                b_all=b_all+b[2*j]
    if flag1==1:
        constraints.append(b_all==1)
pro=cvx.Problem(obj,constraints)
pro.solve(solver=cvx.MOSEK,verbose=True)`

and The error message is：
AttributeError                            Traceback (most recent call last)
<ipython-input-25-5f98dabab458> in <module>()
     80         constraints.append(b_all==1)
     81 pro=cvx.Problem(obj,constraints)
---> 82 pro.solve(solver=cvx.MOSEK,verbose=True)

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\problems\problem.py in solve(self, *args, **kwargs)
    393         else:
    394             solve_func = Problem._solve
--> 395         return solve_func(self, *args, **kwargs)
    396 
    397     @classmethod

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\problems\problem.py in _solve(self, solver, warm_start, verbose, gp, qcp, requires_grad, enforce_dpp, **kwargs)
    700             why the problem could not be solved.
    701         """
--> 702         for parameter in self.parameters():
    703             if parameter.value is None:
    704                 raise error.ParameterError(

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\utilities\performance_utils.py in _compute_once(self, *args, **kwargs)
     68             return cache[key]
     69         else:
---> 70             result = func(self, *args, **kwargs)
     71             cache[key] = result
     72             return result

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\problems\problem.py in parameters(self)
    268         params = self.objective.parameters()
    269         for constr in self.constraints:
--> 270             params += constr.parameters()
    271         return unique_list(params)
    272 

AttributeError: 'bool' object has no attribute 'parameters'
```
I don’t quite understand how this error occurs, like it’s because of the binary number

**Comment by rileyjmurray** [comment]
As you know, in cvxpy we (usually) create constraints with an algebraic syntax like ``expr1 <= expr2``. Sometimes if our code is not set up correctly, we might be in the situation where both ``expr1`` and ``expr2`` are NOT cvxpy objects, and so the python interpreter might evaluate ``expr1 <= expr2`` into a boolean type (rather than a cvxpy Constraint object). The error message you're seeing here is almost always the result of such a mistake in the modeling code. Please look through your code and try to find where you might be accidentally comparing two non-cvxpy objects.

I am closing this issue for now. If you inspect your code in detail and cannot find a mistake of the kind I've suggested above, please leave a comment here again, and I will look into this more.

**Comment by Martinjiajun** [comment]
Hi...in fact I think the error code is as follows, because after I removed the code below, there is no error message like that, but there is a new error message, and I don’t quite understand it.

the removed code is：
for i in range(n_bus):
    b_all=0
    flag1=1
    if root==i:
        for j in range(n_branch):
            if branch[j,0]==i:
                constraints.append(b[j*2]==0)
                flag1=0
            if branch[j,1]==i:
                constraints.append(b[j*2+1]==0)
                flag1=0
    else:
        for j in range(n_branch):
            if branch[j,0]==i:
                b_all=b_all+b[2*j]
            if branch[j,1]==i:
                b_all=b_all+b[2*j+1]
    if flag1==1:
        constraints.append(b_all==1)

And the new error message is：
AssertionError                            Traceback (most recent call last)
<ipython-input-6-b5c6089812f4> in <module>()
     68 
     69 pro=cvx.Problem(obj,constraints)
---> 70 pro.solve(solver=cvx.MOSEK,verbose=True)

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\problems\problem.py in solve(self, *args, **kwargs)
    393         else:
    394             solve_func = Problem._solve
--> 395         return solve_func(self, *args, **kwargs)
    396 
    397     @classmethod

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\problems\problem.py in _solve(self, solver, warm_start, verbose, gp, qcp, requires_grad, enforce_dpp, **kwargs)
    742 
    743         data, solving_chain, inverse_data = self.get_problem_data(
--> 744             solver, gp, enforce_dpp)
    745         solution = solving_chain.solve_via_data(
    746             self, data, warm_start, verbose, kwargs)

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\problems\problem.py in get_problem_data(self, solver, gp, enforce_dpp)
    522             inverse_data = self._cache.inverse_data + [solver_inverse_data]
    523         else:
--> 524             data, inverse_data = solving_chain.apply(self)
    525             safe_to_cache = (
    526                 isinstance(data, dict)

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\reductions\chain.py in apply(self, problem)
     69         inverse_data = []
     70         for r in self.reductions:
---> 71             problem, inv = r.apply(problem)
     72             inverse_data.append(inv)
     73         return problem, inverse_data

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\reductions\complex2real\complex2real.py in apply(self, problem)
     61             # created for the imaginary part.
     62             real_constrs, imag_constrs = self.canonicalize_tree(
---> 63                 constraint, inverse_data.real2imag, leaf_map)
     64             if real_constrs is not None:
     65                 constrs.extend(real_constrs)

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\reductions\complex2real\complex2real.py in canonicalize_tree(self, expr, real2imag, leaf_map)
    133             imag_args = []
    134             for arg in expr.args:
--> 135                 real_arg, imag_arg = self.canonicalize_tree(arg, real2imag, leaf_map)
    136                 real_args.append(real_arg)
    137                 imag_args.append(imag_arg)

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\reductions\complex2real\complex2real.py in canonicalize_tree(self, expr, real2imag, leaf_map)
    133             imag_args = []
    134             for arg in expr.args:
--> 135                 real_arg, imag_arg = self.canonicalize_tree(arg, real2imag, leaf_map)
    136                 real_args.append(real_arg)
    137                 imag_args.append(imag_arg)

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\reductions\complex2real\complex2real.py in canonicalize_tree(self, expr, real2imag, leaf_map)
    133             imag_args = []
    134             for arg in expr.args:
--> 135                 real_arg, imag_arg = self.canonicalize_tree(arg, real2imag, leaf_map)
    136                 real_args.append(real_arg)
    137                 imag_args.append(imag_arg)

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\reductions\complex2real\complex2real.py in canonicalize_tree(self, expr, real2imag, leaf_map)
    133             imag_args = []
    134             for arg in expr.args:
--> 135                 real_arg, imag_arg = self.canonicalize_tree(arg, real2imag, leaf_map)
    136                 real_args.append(real_arg)
    137                 imag_args.append(imag_arg)

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\reductions\complex2real\complex2real.py in canonicalize_tree(self, expr, real2imag, leaf_map)
    138             real_out, imag_out = self.canonicalize_expr(expr, real_args,
    139                                                         imag_args, real2imag,
--> 140                                                         leaf_map)
    141         return real_out, imag_out
    142 

D:\Anaconda\envs\Cvxpy\lib\site-packages\cvxpy\reductions\complex2real\complex2real.py in canonicalize_expr(self, expr, real_args, imag_args, real2imag, leaf_map)
    151             return result
    152         else:
--> 153             assert all(v is None for v in imag_args)
    154             return expr.copy(real_args), None

AssertionError:

**Comment by Martinjiajun** [comment]
> As you know, in cvxpy we (usually) create constraints with an algebraic syntax like `expr1 <= expr2`. Sometimes if our code is not set up correctly, we might be in the situation where both `expr1` and `expr2` are NOT cvxpy objects, and so the python interpreter might evaluate `expr1 <= expr2` into a boolean type (rather than a cvxpy Constraint object). The error message you're seeing here is almost always the result of such a mistake in the modeling code. Please look through your code and try to find where you might be accidentally comparing two non-cvxpy objects.
> 
> I am closing this issue for now. If you inspect your code in detail and cannot find a mistake of the kind I've suggested above, please leave a comment here again, and I will look into this more.

and I test the code ：
a=cvx.Variable(n_branch,boolean=True)
b=cvx.Variable(2*n_branch,boolean=True)
obj=cvx.Maximize(sum(a)+sum(b))
constraints=[]
for i in range(n_bus):
    b_all=0
    flag1=1
    if root==i:
        for j in range(n_branch):
            if branch[j,0]==i:
                constraints.append(b[j*2]==0)
                flag1=0
            if branch[j,1]==i:
                constraints.append(b[j*2+1]==0)
                flag1=0
    else:
        for j in range(n_branch):
            if branch[j,0]==i:
                b_all=b_all+b[2*j]
            if branch[j,1]==i:
                b_all=b_all+b[2*j+1]
    if flag1==1:
        constraints.append(b_all==1)
for i in range(n_branch):
    constraints.append(b[2*i]+b[2*i+1]==a[i])
prob = cvx.Problem(obj,constraints)
prob.solve(solver=cvx.MOSEK,verbose=True)

This integer programming problem can be solved in Cvxpy, which proves that this code can actually be recognized and solved by cvxpy, but an error message will appear when placed in the original problem('bool' object has no attribute 'parameters').This makes me very confused.

**Comment by rileyjmurray** [comment]
I'm reopening the issue because your second error message (the AssertionError) might actually be a bug. However I don't have time to investigate this right now. I will say that I don't quite follow your most recent comment.

A note on formatting: when pasting source code or code output into a GitHub comment, wrap the text in double ticks for in-line formatting (e.g.\`\`simple_expr = single_function_call()\`\`) or tripple ticks for block formatting (e.g. \`\`\` ... code that spans many lines ... \`\`\`) . I've edited your comments so far to have this formatting.

**Comment by Martinjiajun** [comment]
> I'm reopening the issue because your second error message (the AssertionError) might actually be a bug. However I don't have time to investigate this right now. I will say that I don't quite follow your most recent comment.
> 
> A note on formatting: when pasting source code or code output into a GitHub comment, wrap the text in double ticks for in-line formatting (e.g.``simple_expr = single_function_call()``) or tripple ticks for block formatting (e.g. ``` ... code that spans many lines ... ```) . I've edited your comments so far to have this formatting.

My statement may not be clear, and it’s the first time I use github to ask for help. I’m a bit stupid. I actually converted the cvx program in matlab into cvxpy code (I learned from the communication with Professor Boyd last year ). Thank you very much for your reply and help.

**Comment by rileyjmurray** [comment]
I think you're being too hard on yourself. The only thing that's for sure is that you're new to CVXPY. There is bound to be a learning curve with these things. The issue you encountered might actually be a bug, so I will look into it in more detail later on. For now, you can try learning more by using StackOverflow with a CVXPY tag.

**Comment by rileyjmurray** [comment]
Closed by #1551.
